text/0502_aws-vpclattice.md

```diff
@@ -0,0 +1,363 @@
+```
+
+## API Design
+The following API design is proposed
+
+#### Listener
+```typescript
+/**
+ * Create a vpcLattice Listener.
+ * Implemented by `Listener`.
+ */
+export interface IListener extends core.IResource {
+  /**
+  * The Amazon Resource Name (ARN) of the service.
+  */
+  readonly listenerArn: string;
+  /**
+  * The Id of the Service Network
+  */
+  readonly listenerId: string;
+
+  /**
+   * Add A Rule to the Listener
+   */
+  addListenerRule(props: AddRuleProps): void;
+}
+```
+
+#### Service
+```typescript
+/**
+ * Create a vpcLattice service.
+ * Implemented by `Service`.
+ */
+export interface IService extends core.IResource {
+  /**
+  * The Amazon Resource Name (ARN) of the service.
+  */
+  readonly serviceArn: string;
+  /**
+  * The Id of the Service Network
+  */
+  readonly serviceId: string;
+
+  /**
+   * Add An Authentication Policy to the Service.
+   * @param policyStatement[];
+   */
+  addLatticeAuthPolicy(policyStatement: iam.PolicyStatement[]): iam.PolicyDocument;
+  /**
+   * Add A vpc listener to the Service.
+   * @param props
+   */
+  addListener(props: vpclattice.ListenerProps): vpclattice.Listener;
+  /**
+   * Share the service to other accounts via RAM
+   * @param props
+   */
+  share(props: ShareServiceProps): void;
+
+  /**
+  * Create a DNS entry in R53 for the service.
+  */
+  addDNSEntry(props: aws_vpclattice.CfnService.DnsEntryProperty): void;
+
+  /**
+   * Add a certificate to the service
+   * @param certificate
+   */
+  addCertificate(certificate: certificatemanager.Certificate): void;
+
+  /**
+   * add a custom domain to the service
+   * @param domain
+   */
+  addCustomDomain(domain: string): void;
+
+  /**
+   * add a name for the service
+   * @default cloudformation will provide a name
+   */
+  addName(name: string): void;
+  /**
+   *grant a Principal access to a Services Path
+   * @deafult
+   */
+   grantAccess(props: grantAccessProps): void; 
+
+}
+```
+#### ServiceNetwork
+```typescript
+/**
+ * Create a vpc lattice service network.
+ * Implemented by `ServiceNetwork`
+ */
+export interface IServiceNetwork extends core.IResource {
+
+  /**
+  * The Amazon Resource Name (ARN) of the service network.
+  */
+  readonly serviceNetworkArn: string;
+
+  /**
+   * The Id of the Service Network
+   */
+  readonly serviceNetworkId: string;
+  /**
+   * Add LatticeAuthPolicy
+   */
+  addLatticeAuthPolicy(policyDocument: iam.PolicyStatement[]): void;
+  /**
+   * Add Lattice Service Policy
+   */
+  addService(service: vpclattice.Service): void;
+  /**
+   * Associate a VPC with the Service Network
+   */
+  associateVPC(vpc: ec2.Vpc, securityGroups: ec2.SecurityGroup[]): void;
+  /**
+   * Log To S3
+   */
+  logToS3(bucket: s3.Bucket | s3.IBucket ): void;
+  /**
+   * Send Events to Cloud Watch
+   */
+  sendToCloudWatch(log: logs.LogGroup | logs.ILogGroup ): void;
+  /**
+   * Stream to Kinesis
+   */
+  streamToKinesis(stream: kinesis.Stream | kinesis.IStream ): void;
```

Comment:
jsii does not currently support union types.

Reply:
But these can all just take in the interface.

Reply:
This is now resolved, no unions are used anywhere in the constructs.. The use of unions is a bit vauge in the jsii docs, its not entirely a hard and fast rule it seems, however, it seems best to avoid it if we can, and I have.